streams/src/main/java/org/apache/kafka/streams/kstream/KStream.java

```diff
@@ -406,6 +406,20 @@ void writeAsText(final String filePath,
     void foreach(final ForeachAction<? super K, ? super V> action);
 
     /**
+     * Perform an action on each record of {@code KStream}.
+     * This is a stateless record-by-record operation (cf. {@link #process(ProcessorSupplier, String...)}).
+     *
+     * Peek is a non-terminal operation that triggers a side effect (such as logging or statistics collection)
+     * and returns an unchanged stream.
```

Comment:
@miguno Actually, we do not enforce this guarantee. For many UDFs, we just hand in a reference to key and/or value and if the user would alter the object, we would not notice...